### DISCUSSION TITLE
Supplying user-specified gradient

### DISCUSSION THREAD

**Discussion by vialvirt** [scikit-hep/iminuit#926] [Q&A]
Hello to iminuit community and sages!

I am studying how to supply Imiuit with a user-specified gradient function.  The [example of using automatic differentiation](https://iminuit.readthedocs.io/en/stable/notebooks/automatic_differentiation.html#Automatic-differentiation-with-JAX) gave pointers on what is possible, however I would like to provide the minimizer with a gradient function directly and seem to have trouble understanding what form iminuit needs it in, as the snippet runs into errors.  Here is bit of example code to reproduce the situation: 

    import numpy as np 
    from numpy import random
    from iminuit import cost, Minuit
    
    
    random.seed(1)
    
    
    
    # generate data
    
    loc = 0.0
    scale = 0.1
    size = 100
    
    data = random.normal(loc=loc, scale=scale, size=size)
    
    
    # fitting function & gradient
    
    def gaussian(x,mu,sigma,A):
        
        g = A/(sigma*np.sqrt(2*np.pi))*np.exp(-(x-mu)**2/(2*sigma**2))
        
        return A, g    
    
    def gradient(x,mu,sigma,A):
        
        g = gaussian(x,mu,sigma,A)[1]
        
        dfdx  = g * (mu-x)/sigma**2
        dfdmu = g * (x-mu)/sigma**2
        dfds  = g * ((x-mu)**2/sigma**3 - A/(sigma**2*np.sqrt(2*np.pi)))
        dfdA  = g / A 
        
        return (dfdx,dfdmu,dfds,dfdA)
    
    
    d = cost.ExtendedUnbinnedNLL(data, gaussian,log = False)
    m = Minuit(d,mu = 0,sigma = 0.1,A = 100,grad = gradient, name = ("mu","sigma","A"))
    
    m.migrad()
    print(m)

 

With this an error 
```

line 694, in migrad
    fm = migrad(ncall, self._tolerance)

TypeError: gradient() missing 1 required positional argument: 'A'
```

appeared. I thought that perhaps the gradient would need to be in form gradient(x,*params) -> gradient(*params)
Trying  this option produced the following error:

```
line 694, in migrad
    fm = migrad(ncall, self._tolerance)

RuntimeError: Unable to cast Python instance to C++ type (compile in debug mode for details) 
```

There must be something that I am missing here but can't quite pick up on.  How would one need to go about supplying the gradient to iminuit?

**Comment by HDembinski** [comment]
The builtin cost functions like ExtendedUnbinnedNLL currently do not support models with gradients. This should be added as a feature. If you want to use gradients at the moment, you need to write the negative log-likelihood function yourself and compute the gradient of that.

What you were trying is to pass the gradient of the model (the PDF) to Minuit, but you need to pass the gradient of the ExtendedUnbinnedNLL to iminuit.

In the future, the builtin cost functions should allow passing a gradient of the model and then pass on the gradient of the cost function to iminuit automatically. But this is not implemented yet.

Supporting gradients does not have high priority, since the studies showed that a precomputed gradient does not provide a large benefit. Internally, Minuit's Migrad algorithm is also not well prepared to handle user-provided gradients.
